project_cost_report/report.py

```diff
@@ -0,0 +1,436 @@
+        """Get the stockable/consumable product categories.
+
+        :param project: the project.project record
+        :param report_context: the rendering context
+        :rtype: dict
+        """
+        domain = self._get_product_analytic_line_domain(project)
+        lines = self.env['account.analytic.line'].search(domain)
+        grouped_lines = _group_analytic_lines(lines, lambda l: l.product_id.categ_id)
+        sorted_categories = sorted(grouped_lines.keys(), key=lambda c: c.name)
+        unfolded_categories = report_context.get('unfolded_categories') or {}
+        unfolded_product_categories = unfolded_categories.get('product') or []
+        return [
+            CostReportCategory(
+                id_=c.id,
+                name=c.name,
+                lines=grouped_lines.get(c),
+                folded=c.id not in unfolded_product_categories
+            ) for c in sorted_categories
+        ]
+
+    def _get_product_total(self, project):
+        """Get the total amount for the PRODUCTS section.
+
+        :param project: the project.project record
+        :rtype: float
+        """
+        domain = self._get_product_analytic_line_domain(project)
+        result = self.env['account.analytic.line'].read_group(
+            domain=domain, fields=['account_id', 'amount'], groupby='account_id')
+        amount = result[0]['amount'] if result else 0
+        return float_round(amount, 2)
+
+    def _get_rendering_variables(self, project, report_context):
+        """Add the variables related to the PRODUCTS section."""
+        res = super()._get_rendering_variables(project, report_context)
+        res.update({
+            'product_categories': self._get_product_categories(project, report_context),
+            'product_total': self._get_product_total(project),
+        })
+        return res
+
+    @api.model
+    def get_foldable_categories(self, project_id):
+        """Add product categories to foldable categories."""
+        res = super().get_foldable_categories(project_id)
+        project = self.env['project.project'].browse(project_id)
+        domain = self._get_product_analytic_line_domain(project)
+        analytic_lines = self.env['account.analytic.line'].search(domain)
+        res['product'] = analytic_lines.mapped('product_id.categ_id.id')
+        return res
+
+
+class TimeCategory(CostReportCategory):
+    """Categories used for the TIME section.
+
+    These categories have one more field to display: the total hours.
+
+    The total of units can not be used on other sections because
+    other sections mix different units of measure.
+    """
+
+    def __init__(self, id_: int, name: str, lines: models.Model, folded: bool):
+        super().__init__(id_, name, lines, folded)
+        self.total_hours = float_round(sum(l.unit_amount for l in lines), 2)
+
+
+class ProjectCostReportWithTime(models.TransientModel):
+
+    _inherit = 'project.cost.report'
+
+    def _get_empty_task_type_label(self):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=3675949)